core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+                        tx not in old_state)]
+        return tx_diffs
+
+    def get_state_diffs(self, event_filter: object) -> list:
+        """
+        Gets state, then finds diffs, then sets state of blockchain.
+        """
+        new_state = self.get_global_state()
+        if new_state:
+            state_diffs = self.get_diffs(self.state, new_state)
+            filtered_diffs = [tx for tx in state_diffs if event_filter(tx)]
+            logging.info("filtered diffs:{}".format(filtered_diffs))
+            self.update_state(new_state)
+            return filtered_diffs
+
+    def check_malformed_content(self) -> None:
+        """
+        Checks whether content is of the form
+            `{
+                type: ... ,
+                payload: ...
+            }`
+        """
+        pass
+
+    def update_state(self, new_state_wrapper: list) -> None:
+        """
+        Given the freshly-downloaded state, call a handler on each transaction
+        that was not already present in our own state
+        """
+        new_state = dict(new_state_wrapper).get('messages')
+        len_state = len(self.state)
+        for i in new_state[len_state:]:
+            self.state.append(i)
+
+    def setter(self, key: str, value: object, flag: bool = False) -> str:
+        """
+        Provided a key and a JSON/np.array object, upload the object to
+        IPFS and then store the hash as the value on the blockchain. The key
+        should be a backward reference to a prior tx
+        """
+        on_chain_value = self._upload(value) if value else None
+        key = on_chain_value if flag else key
+        tx = {'key': key, 'content': on_chain_value}
+        try:
+            tx_receipt = requests.post(
+                "http://localhost:{0}/txs".format(self.port), json=tx)
+            tx_receipt.raise_for_status()
+        except Exception as e:
+            logging.info("HTTP POST error, got: {0}".format(e))
+        return tx_receipt.text
+
+    def getter(self, key: str) -> list:
+        """
+        Next, provided a key, get the IPFS hash
+        from the blockchain and download the object from IPFS
+        """
+        logging.info("Getting from blockchain...")
+        self.update_state(self.get_global_state())
+        retval = self._download(key)
+        return retval
+
+    def _upload(self, obj: object) -> str:
+        """
+        Provided any Python object, store it on IPFS and then upload
+        the hash that will be uploaded to the blockchain as a value
+        """
+        ipfs_hash = self._content_to_ipfs(obj)
+        return str(ipfs_hash)
+
+    def _download(self, key: str) -> object:
+        """
+        Provided an on-chain key, retrieve the value from local state and
+        retrieve the Python object from IPFS
+        TODO: implement a better way to parse through state list
+        """
+        relevant_txs = [self._ipfs_to_content(tx.get('content'))
```

Comment:
This is traversing the list from index 0 to N. Why not from N to 0? (I'm assuming things are appended to the list instead of prepended.)

Also, what if the 2 `.get()`'s fail?

Reply:
can you elaborate on 'why not from N to 0?'
also .get()'s shouldn't fail

Reply:
@georgymh would `assert` statements here help with the gets?

Reply:
Yes @nzoghb 

@kiddyboots216, what I mean is why are we traversing from the beginning of time instead of the end of time backwards?